examples/jwt-auth/src/jwt.rs

```diff
@@ -0,0 +1,769 @@
+                }
+            };
+            hmac = Some(JwtHmac {
+                algorithm,
+                verifier,
+            });
+        }
+        hmac
+    }
+}
+
+#[derive(Debug, Default, Deserialize, JsonSchema)]
+struct Conf {
+    algorithm: String,
+    key: String,
+    time_tolerance: Option<u64>,
+    max_token_life: Option<u64>,
+}
+
+#[async_trait::async_trait]
+impl Plugin for JwtAuth {
+    type Config = Conf;
+
+    async fn startup(&mut self) -> Result<(), BoxError> {
+        tracing::debug!("starting: {}: {}", stringify!(JwtAuth), self.name());
+        Ok(())
+    }
+
+    async fn shutdown(&mut self) -> Result<(), BoxError> {
+        tracing::debug!("shutting down: {}: {}", stringify!(JwtAuth), self.name());
+        Ok(())
+    }
+
+    fn router_service(
+        &mut self,
+        service: BoxService<RouterRequest, RouterResponse, BoxError>,
+    ) -> BoxService<RouterRequest, RouterResponse, BoxError> {
+        // We are going to use the `jwt-simple` crate for our JWT verification.
+        // The crate provides straightforward support for the popular JWT algorithms.
+
+        // `ServiceBuilder` provides us with an `checkpoint` method.
+        //
+        // This method allows us to return ControlFlow::Continue(request) if we want to let the request through,
+        // or ControlFlow::Break(response) with a crafted response if we don't want the request to go through.
+
+        // Clone/Copy the data we need in our closure.
+        let mut hmac_verifier = None;
+        if let Some(hmac) = &self.hmac {
+            hmac_verifier = Some(hmac.verifier.clone());
+        }
+        let time_tolerance = self.time_tolerance;
+        let max_token_life = self.max_token_life;
+
+        ServiceBuilder::new()
+            .checkpoint(move |req: RouterRequest| {
+                // We are going to do a lot of similar checking so let's define a local function
+                // to help reduce repetition
+                fn failure_message(
+                    msg: String,
+                    status: StatusCode,
+                ) -> Result<ControlFlow<RouterResponse, RouterRequest>, BoxError> {
+                    let res = plugin_utils::RouterResponse::builder()
+                        .errors(vec![apollo_router_core::Error {
+                            message: msg,
+                            ..Default::default()
+                        }])
+                        .build()
+                        .with_status(status);
+                    Ok(ControlFlow::Break(res))
+                }
+
+                // The http_request is stored in a `RouterRequest` context.
+                // We are going to check the headers for the presence of the header we're looking for
+                // We are implementing: https://www.rfc-editor.org/rfc/rfc6750
+                // so check for our AUTHORIZATION header.
+                if !req.context.request.headers().contains_key(AUTHORIZATION) {
```

Comment:
why not do this check and the next in one? e.g.

```rust
let jwt_value_result = match req
                    .context
                    .request
                    .headers()
                    .get(AUTHORIZATION) {
    None => 
    // Prepare an HTTP 401 response with a GraphQL error message
    return failure_message(
        format!("Missing '{}' header", AUTHORIZATION),
        StatusCode::UNAUTHORIZED,
    ),
    Some(value) => value.to_str(),
};
```

Reply:
Done: 235fd34435a22b3df5d74473c50fb3be5dd96ffa